cleverhans/model.py

```diff
@@ -0,0 +1,221 @@
+from abc import ABCMeta
+
+
+class Model(object):
+    """
+    An abstract interface for model wrappers that exposes model symbols
+    needed for making an attack. This abstraction removes the dependency on
+    any specific neural network package (e.g. Keras) from the core
+    code of CleverHans. It can also simplify exposing the hidden features of a
+    model when a specific package does not directly expose them. For an
+    example of why this is useful, see the paper "Adversarial Manipulation
+    of Deep Representations" (https://arxiv.org/abs/1511.05122).
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
+        """
+        Init a wrapper. If `fprop_layer` is implemented, `__init__`
+        should keep track of the name of the layers or `self.model` should
+        provide a method for retrieving a layer.
+        :param model: A function that takes a symbolic input and returns the
+                      symbolic output for the model's post-softmax predictions
+                      (probabilities).
+        """
+        self.model = model
+
+        # The following is a cache to prevent the construction of increasingly
+        # large graphs after multiple calls of the fprop methods. The cache is
+        # implemented as a dictionary of the form (input, train): output_dict
+        # The key is a pair of input (the symbolic representation of the input)
+        # and train (a boolean indicating whether the graph is in training or
+        # inference mode---which changes the behavior of layers like dropout).
```

Comment:
I'm not a big fan of the boolean for controlling behavior.

There are times when you have more than two behaviors. There are times when you want Monte Carlo dropout at test time.

It's probably better to get different behaviors by requesting different endpoints or by feeding different values to placeholders that control behavior at different layers.